File: src/main/webapp/js/faq.js
// FAQ 화면 공통 스크립트 (강선구)
// faq_list.jsp / faq_write.jsp / faq_update.jsp 가 같이 쓴다.
// 세 화면 모두 <form name="faq" action="Faq"> 를 갖고 있어서, t_gubun 만 채우고 submit 하면
// Faq 서블릿이 분기한다 (팀 공통 common.js 의 movePage 와 같은 방식).

// 빈 값 검사. 비어 있으면 alert 띄우고 그 칸에 커서를 옮긴 뒤 true 를 돌려준다.
function isEmpty(obj, msg) {
	if (obj.value.trim() == "") {
		alert(msg);
		obj.focus();
		return true;
	}
	return false;
}

// 목록 → 수정 화면
function goUpdateForm(faq_id) {
	var form = document.faq;
	form.t_gubun.value = "updateForm";
	form.t_faq_id.value = faq_id;
	form.submit();
}

// 목록 → 삭제. 되돌릴 수 없으니 confirm 을 한 번 거친다
function goDelete(faq_id) {
	if (!confirm("정말 삭제하시겠습니까?\n삭제한 글은 되돌릴 수 없습니다.\n(잠시 감추기만 하려면 [수정] → 노출여부 '숨김')")) return;
	var form = document.faq;
	form.t_gubun.value = "delete";
	form.t_faq_id.value = faq_id;
	form.submit();
}

// 등록 화면 [저장]
function goSave() {
	var form = document.faq;
	if (isEmpty(form.t_question, "질문을 입력하세요.")) return;
	if (isEmpty(form.t_answer,   "답변을 입력하세요.")) return;
	form.t_gubun.value = "save";
	form.submit();
}

// 수정 화면 [수정완료]
function goUpdate() {
	var form = document.faq;
	if (isEmpty(form.t_question, "질문을 입력하세요.")) return;
	if (isEmpty(form.t_answer,   "답변을 입력하세요.")) return;
	form.t_gubun.value = "update";
	form.submit();
}

File: src/main/java/controller/Faq.java
package controller;

import java.io.IOException;
import java.util.ArrayList;

import javax.servlet.RequestDispatcher;
import javax.servlet.ServletException;
import javax.servlet.annotation.WebServlet;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

import command.faq.FaqDelete;
import command.faq.FaqSave;
import command.faq.FaqUpdate;
import common.CommonExecute;
import dao.FaqDao;
import dto.FaqDto;


@WebServlet("/Faq")
public class Faq extends HttpServlet {
	private static final long serialVersionUID = 1L;

	@Override
	protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
		request.setCharacterEncoding("UTF-8");
		String gubun = request.getParameter("t_gubun");
		if (gubun == null) gubun = "list";

		// 목록 말고는 전부 관리자 전용. 화면에서 버튼을 숨겨도 주소를 직접 칠 수 있으니 서버에서 막는다.
		boolean adminOnly = gubun.equals("writeForm") || gubun.equals("save")
				|| gubun.equals("updateForm") || gubun.equals("update")
				|| gubun.equals("delete");

		if (adminOnly && !isAdmin(request)) {
			request.setAttribute("t_msg", "관리자만 이용할 수 있는 기능입니다.");
			request.setAttribute("t_url", "Faq");
			forward(request, response, "common_alert.jsp");
			return;
		}

		if (gubun.equals("writeForm")) {
			forward(request, response, "faq/faq_write.jsp");
			return;
		} else if (gubun.equals("save")) {
			CommonExecute cmd = new FaqSave();
			cmd.execute(request);
			forward(request, response, "common_alert.jsp");
			return;
		} else if (gubun.equals("updateForm")) {
			int faq_id = parseId(request.getParameter("t_faq_id"));
			FaqDto dto = new FaqDao().getFaqView(faq_id);
			if (dto == null) {
				request.setAttribute("t_msg", "존재하지 않는 글입니다.");
				request.setAttribute("t_url", "Faq");
				forward(request, response, "common_alert.jsp");
				return;
			}
			request.setAttribute("dto", dto);
			forward(request, response, "faq/faq_update.jsp");
			return;
		} else if (gubun.equals("update")) {
			CommonExecute cmd = new FaqUpdate();
			cmd.execute(request);
			forward(request, response, "common_alert.jsp");
			return;
		} else if (gubun.equals("delete")) {
			CommonExecute cmd = new FaqDelete();
			cmd.execute(request);
			forward(request, response, "common_alert.jsp");
			return;
		}

		String category = request.getParameter("t_category");
		if (category == null) category = "";
		boolean isAdmin = isAdmin(request);

		ArrayList<FaqDto> dtos = new FaqDao().getFaqList(category, isAdmin);

		request.setAttribute("dtos", dtos);
		request.setAttribute("category", category);
		request.setAttribute("isAdmin", isAdmin);
		forward(request, response, "faq/faq_list.jsp");
	}

	private boolean isAdmin(HttpServletRequest request) {
		return "top".equals(request.getSession().getAttribute("sessionLevel"));
	}

	public static int parseId(String s) {
		if (s != null && s.matches("[0-9]+")) return Integer.parseInt(s);
		return 0;
	}

	private void forward(HttpServletRequest request, HttpServletResponse response, String page)
			throws ServletException, IOException {
		RequestDispatcher rd = request.getRequestDispatcher(page);
		rd.forward(request, response);
	}

	@Override
	protected void doPost(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
		doGet(request, response);
	}

}
